File: funtime/pylib/simon_rules.py
"""simon_rules.py - the rules of Simon Says, in Python.

Four pads. The game flashes a sequence, then you play it back; every round it
grows by one and the flashes get quicker.

The game is always in one of three PHASES: "watch", "play" or "over".
"""

import random

PAD_COUNT = 4


def random_pad():
    """Pick one of the four pads (0 to 3)."""
    return random.randrange(PAD_COUNT)


def add_step(sequence):
    """Make the sequence one longer.

    INPUT:  sequence. OUTPUT: a NEW list with one more random pad on the end.
    """
    return list(sequence) + [random_pad()]


def is_correct_so_far(sequence, input_pads):
    """Is the player's answer still right?

    INPUT:  sequence - what the game flashed. input_pads - what was pressed.
    OUTPUT: True while every pad pressed matches the sequence.
    ALGORITHM: the input cannot be longer than the sequence, and every pad
    must match the one in the same place. An empty input is fine.
    """
    if len(input_pads) > len(sequence):
        return False
    return all(pad == sequence[i] for i, pad in enumerate(input_pads))


def is_round_complete(sequence, input_pads):
    """Has the player repeated the whole sequence?"""
    return is_correct_so_far(sequence, input_pads) and len(input_pads) == len(sequence)


def flash_interval(round_number):
    """How long each flash lasts, in milliseconds.

    ALGORITHM: 620 ms in round 1, 25 ms quicker each round, never below 260.
    """
    return max(260, 620 - (round_number - 1) * 25)


def score_for_round(round_number):
    """How many points is finishing a round worth? 10 per pad remembered."""
    return 10 * round_number


def create_game():
    """Start a brand-new game, already flashing the first pad."""
    sequence = add_step([])
    return {
        "sequence": sequence,
        "input": [],
        "round": 1,
        "score": 0,
        "phase": "watch",
        "flash_index": 0,
        "flash_on": True,
        "flash_timer": 0,
        "lit": sequence[0],
        "is_over": False,
        "is_paused": False,
    }


def press_pad(state, pad):
    """The player presses one pad.

    ALGORITHM: only during the player's turn; add the pad, light it, end the
    game if it is wrong, and start the next round if it completes the sequence.
    """
    if state["phase"] != "play" or state["is_over"] or state["is_paused"]:
        return False

    state["input"].append(pad)
    state["lit"] = pad
    state["flash_timer"] = 0

    if not is_correct_so_far(state["sequence"], state["input"]):
        state["phase"] = "over"
        state["is_over"] = True
        return True

    if is_round_complete(state["sequence"], state["input"]):
        state["score"] += score_for_round(state["round"])
        state["round"] += 1
        state["sequence"] = add_step(state["sequence"])
        state["input"] = []
        state["phase"] = "watch"
        state["flash_index"] = 0
        state["flash_on"] = True
        state["lit"] = state["sequence"][0]
    return True


def update_game(state, elapsed_ms):
    """Let time pass (called about 60 times a second).

    ALGORITHM: while the game is flashing, count the timer up; each time it
    passes this round's interval, either turn the pad off (a gap) or move on to
    the next pad. After the last flash it becomes the player's turn.
    """
    if state["is_over"] or state["is_paused"]:
        return

    if state["phase"] == "watch":
        state["flash_timer"] += elapsed_ms
        if state["flash_timer"] >= flash_interval(state["round"]):
            state["flash_timer"] = 0
            if state["flash_on"]:
                state["flash_on"] = False
                state["lit"] = -1
            else:
                state["flash_on"] = True
                state["flash_index"] += 1
                if state["flash_index"] >= len(state["sequence"]):
                    state["phase"] = "play"
                    state["lit"] = -1
                else:
                    state["lit"] = state["sequence"][state["flash_index"]]
        if state["phase"] == "watch" and state["flash_on"]:
            state["lit"] = state["sequence"][state["flash_index"]]
    elif state["lit"] != -1:
        state["flash_timer"] += elapsed_ms
        if state["flash_timer"] > 220:
            state["lit"] = -1
            state["flash_timer"] = 0


def toggle_pause(state):
    """Freeze or unfreeze the game."""
    if not state["is_over"]:
        state["is_paused"] = not state["is_paused"]


def action_for_key(key):
    """Turn a keyboard key into an action name, or None."""
    keys = {
        "1": "pad0", "q": "pad0", "arrowleft": "pad0",
        "2": "pad1", "w": "pad1", "arrowup": "pad1",
        "3": "pad2", "a": "pad2", "arrowdown": "pad2",
        "4": "pad3", "s": "pad3", "arrowright": "pad3",
        "p": "pause",
        "r": "restart",
    }
    return keys.get(str(key).lower())

File: funtime/lib/simon-rules.js
/* ============================================================
   simon-rules.js — the rules of Simon Says

   Four pads. The game flashes a sequence, then you play it back. Every round
   the sequence grows by one, so it is really a test of your memory — and the
   flashes get quicker as you go.

   The game is always in one of three PHASES:
       'watch'  — the game is flashing the sequence at you
       'play'   — your turn to repeat it
       'over'   — you got one wrong
   ============================================================ */

const PAD_COUNT = 4;

/**
 * randomPad — pick one of the four pads.
 * INPUT: none. OUTPUT: 0, 1, 2 or 3.
 */
function randomPad() {
    return Math.floor(Math.random() * PAD_COUNT);
}

/**
 * addStep — make the sequence one longer.
 *
 * INPUT:  sequence — the pads so far
 * OUTPUT: a NEW list with one more random pad on the end
 *
 * ALGORITHM: copy the list and push a random pad onto it. The old sequence is
 *            left alone, which makes this easy to test.
 */
function addStep(sequence) {
    const longer = sequence.slice();
    longer.push(randomPad());
    return longer;
}

/**
 * isCorrectSoFar — is the player's answer still right?
 *
 * INPUT:  sequence — what the game flashed. input — what the player has pressed.
 * OUTPUT: true while every pad pressed so far matches the sequence
 *
 * ALGORITHM: the input must not be longer than the sequence, and every pad in
 *            it must match the pad at the same place. An empty input is fine —
 *            the player has not gone wrong yet.
 */
function isCorrectSoFar(sequence, input) {
    if (input.length > sequence.length) {
        return false;
    }
    for (let i = 0; i < input.length; i++) {
        if (input[i] !== sequence[i]) {
            return false;
        }
    }
    return true;
}

/**
 * isRoundComplete — has the player repeated the whole sequence?
 * INPUT: sequence, input. OUTPUT: true when the input matches it exactly.
 * ALGORITHM: it must be right so far AND the same length.
 */
function isRoundComplete(sequence, input) {
    return isCorrectSoFar(sequence, input) && input.length === sequence.length;
}

/**
 * flashInterval — how long each flash lasts, in milliseconds.
 *
 * INPUT:  round — which round you are on (1, 2, 3, …)
 * OUTPUT: milliseconds
 *
 * ALGORITHM: 620 ms in round 1, 25 ms quicker each round, never below 260 —
 *            fast enough to be exciting, slow enough to follow.
 */
function flashInterval(round) {
    const milliseconds = 620 - (round - 1) * 25;
    if (milliseconds < 260) {
        return 260;
    }
    return milliseconds;
}

/**
 * scoreForRound — how many points is finishing a round worth?
 * INPUT: round. OUTPUT: the points to add.
 * ALGORITHM: 10 points per pad you had to remember, so later rounds pay more.
 */
function scoreForRound(round) {
    return 10 * round;
}

/** createGame — start a brand-new game, already flashing the first pad. */
function createGame() {
    return {
        sequence: addStep([]),
        input: [],
        round: 1,
        score: 0,
        phase: 'watch',
        flashIndex: 0,
        flashOn: true,
        flashTimer: 0,
        lit: -1,
        isOver: false,
        isPaused: false
    };
}

/**
 * pressPad — the player presses one pad.
 *
 * INPUT:  state — the game. pad — 0 to 3.
 * OUTPUT: true if the press counted
 *
 * ALGORITHM:
 *   1. Ignore presses unless it is the player's turn.
 *   2. Add the pad to the input and light it up.
 *   3. If the answer has gone wrong, the game is over.
 *   4. If the round is complete, score it, grow the sequence and start
 *      flashing again.
 */
function pressPad(state, pad) {
    if (state.phase !== 'play' || state.isOver || state.isPaused) {
        return false;
    }

    state.input.push(pad);
    state.lit = pad;
    state.flashTimer = 0;

    if (!isCorrectSoFar(state.sequence, state.input)) {
        state.phase = 'over';
        state.isOver = true;
        return true;
    }

    if (isRoundComplete(state.sequence, state.input)) {
        state.score = state.score + scoreForRound(state.round);
        state.round = state.round + 1;
        state.sequence = addStep(state.sequence);
        state.input = [];
        state.phase = 'watch';
        state.flashIndex = 0;
        state.flashOn = true;
        state.lit = state.sequence[0];
    }
    return true;
}

/**
 * updateGame — let time pass (called about 60 times a second).
 *
 * INPUT:  state, elapsedMs
 * OUTPUT: nothing
 *
 * ALGORITHM: while the game is flashing, count the timer up; each time it
 *            passes this round's interval, either turn the pad off (a gap) or
 *            move on to the next pad. When the last flash is done it becomes
 *            the player's turn.
 */
function updateGame(state, elapsedMs) {
    if (state.isOver || state.isPaused) {
        return;
    }

    if (state.phase === 'watch') {
        state.flashTimer = state.flashTimer + elapsedMs;
        const interval = flashInterval(state.round);
        if (state.flashTimer >= interval) {
            state.flashTimer = 0;
            if (state.flashOn) {
                state.flashOn = false;
                state.lit = -1;
            } else {
                state.flashOn = true;
                state.flashIndex = state.flashIndex + 1;
                if (state.flashIndex >= state.sequence.length) {
                    state.phase = 'play';
                    state.lit = -1;
                } else {
                    state.lit = state.sequence[state.flashIndex];
                }
            }
        }
        if (state.phase === 'watch' && state.flashOn) {
            state.lit = state.sequence[state.flashIndex];
        }
    } else if (state.lit !== -1) {
        state.flashTimer = state.flashTimer + elapsedMs;
        if (state.flashTimer > 220) {
            state.lit = -1;
            state.flashTimer = 0;
        }
    }
}

/** togglePause — freeze or unfreeze the game. */
function togglePause(state) {
    if (!state.isOver) { state.isPaused = !state.isPaused; }
}

/**
 * actionForKey — turn a keyboard key into an action name, or null.
 * The four pads are also the keys 1, 2, 3 and 4.
 */
function actionForKey(key) {
    const k = String(key).toLowerCase();

    if (k === '1' || k === 'q' || k === 'arrowleft') { return 'pad0'; }
    if (k === '2' || k === 'w' || k === 'arrowup') { return 'pad1'; }
    if (k === '3' || k === 'a' || k === 'arrowdown') { return 'pad2'; }
    if (k === '4' || k === 's' || k === 'arrowright') { return 'pad3'; }
    if (k === 'p') { return 'pause'; }
    if (k === 'r') { return 'restart'; }

    return null;
}
